### PR TITLE
Use modern type annotations

### PR DESCRIPTION
**Pull request by hhoppe**
(No functional changes.)

Use modern annotations (using `from __future__ import annotations`) and types in `collections.abc`.

In `pyproject.toml`, added configurations for `autopep8`, `mypy`, and `pylint` so that these tools report no errors.
```
echo autopep8; autopep8 -j8 -d .; echo mypy; mypy .; echo pylint; pylint -j8 .; echo pytest; pytest -qq
```

I verified that all works fine in Colab using
`!pip install git+https://github.com/hhoppe/mediapy.git@annotations` in several notebooks.

My next step is to update the online API documentation, to use `pdoc` rather than `pdoc3`, and to use an automated GitHub workflow action.

### CODE REVIEW

**Comment by Conchylicultor** [comment]
Thank you, unfortunately Google is still running under Python 3.9. So `pytype` trigger internally:

```
unsupported operand type(s) for |: '_Path: Type[Union[str, os.PathLike]]' and 'None: None'
```

We're planning to switch to 3.10 around January 15th. After which I think this could be submitted.

However be aware this would trigger error if users try to explicitly access the typing annotations in 3.8 and 3.9:

```python
def f() -> int | str:
  pass

typing_extensions.get_type_hints(f)
```

Raise in 3.9 and 3.8

```
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

But it's unlikely that user try `typing_extensions.get_type_hints` on mediapy API, so we can still submit this after Google use 3.10 by default

File: pyproject.toml [hunk 1]
```diff
@@ -45,3 +45,35 @@ dev = [
 [build-system]
 requires = ["flit_core >=3.5,<4"]
 build-backend = "flit_core.buildapi"
+
+[tool.autopep8]
```
Comment (Conchylicultor) [inline pyproject.toml:49]: (Maybe not in this cl) Can we use `pyink` instead ? This is becoming the Google default, a fork of black adapted to Google style. See: https://github.com/google/pyink

```toml
[tool.pyink]
# Formatting configuration to follow Google style-guide
pyink-indentation = 2
pyink-use-majority-quotes = true
```

And for vscode compatibility: https://github.com/google/pyink#is-there-a-vs-code-extension-for-pyink

File: pyproject.toml [hunk 2]
```diff
@@ -45,3 +45,35 @@ dev = [
...
+
+[tool.mypy]
+strict = true
+ignore_missing_imports = true
+
+[[tool.mypy.overrides]]
+module = "mediapy_test"
+ignore_errors = true
+
+[tool.pylint.main]
```
Comment (Conchylicultor) [inline pyproject.toml:64]: (Just for info) Alternatively there's a google-specific `.pylintrc`: https://github.com/google-research/visu3d/blob/main/.pylintrc

**Comment by hhoppe** [comment]
Thanks!
January 15 is not far at all --- keeping fingers crossed..
I spent some time today looking at "pyink"; it seems very nice.

**Comment by Conchylicultor** [comment]
For pyink, see https://github.com/google-research/visu3d/pull/89/files as how I setup it in my other projects

**Comment by Conchylicultor** [comment]
Our tests now pass at Google. I'll still wait a few days before submitting this in case 3.10 is rolled back internally

**Comment by Conchylicultor** [comment]
Submitting this now. Thank you for your patience!

**Comment by Conchylicultor** [comment]
Closing this as this was merged in https://github.com/google/mediapy/commit/5270f0e8d180721cb52e8c43955cb1754ad11bdf. Not sure why copybara hasn't closed this

**Pull request closed without merging**